### ISSUE TITLE
Can't write files if using HDF5_PLUGIN_PATH .

### ISSUE THREAD

**Issue by orioltinto** [silx-kit/hdf5plugin#204]
Hello,

because it would be great to use the filters outside python as well, I was trying to use them by setting the environment variable **HDF5_PLUGIN_PATH**.

If I set this variable and I try to read a compressed file without loading hdf5plugin it works without problems (**GOOD!**).

However, when I try to use the filters to write a file it raises an **error from h5py**:

```python
  File "/usr/local/lib/python3.10/dist-packages/h5py/_hl/group.py", line 161, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "/usr/local/lib/python3.10/dist-packages/h5py/_hl/dataset.py", line 156, in make_new_dset
    dset_id = h5d.create(parent.id, name, tid, sid, dcpl=dcpl, dapl=dapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5d.pyx", line 87, in h5py.h5d.create
ValueError: Unable to create dataset (error during user callback)
```

Also, if the environment variable is set, when importing **hdf5plugin** the libraries are **not initialized** so it **keeps failing**.

In summary, we **need to define** HDF5_PLUGIN_PATH to allow other tools to use the filters but it makes hdf5plugin **fail**.

**Comment by t20100** [comment]
Hi,

Yes you can only read when setting the `hdf5plugin` filters through `HDF5_PLUGIN_PATH`.

The problem is that for writing, some filters need to call some functions of `libhdf5` to get e.g., the data type or the item size, and so you need to link the filter with libhdf5.... This is an issue because the `libhdf5` that one use at runtime through a `h5py` wheel for example is not known at compile time of `hdf5plugin` + C symbols for Python extension are not loaded in the global namespace so not accessible to other packages.

To workaround this `hdf5plugin` provides stubs of `libhdf5` that are initialized at runtime and dynamically load C symbols from the `libhdf5` used by `h5py`.

For the issue of defining  `HDF5_PLUGIN_PATH` making it not loading the filters, it's indeed not good. We never had this use case so far, it was either `HDF5_PLUGIN_PATH` or `h5py`.
I'm not sure we can get the information from `libhdf5` of where each already loaded filters comes from and detect that it is one from `hdf5plugin` loaded through `HDF5_PLUGIN_PATH`. This has to be checked.
Also, we'll have to check the unregister function and if what is the priority among filters with same IDs.

If we can't detect from where the filters comes from, we can anyway try to provide a function like `register_filters(filters=('sz', 'blosc'), force=True)` which allows to override already loaded filters.
Would that work in your case?

**Comment by orioltinto** [comment]
Hi @t20100,
ok I understand.
I guess that mostly we would use non-python tools like CDO to read the files, not writing them.
However, at some point we will use the filters from Fortran/C++ scientific models, but I guess that we can compile the filters separately along with the models whenever this happens.

At some point we had this workaround in our code to unset **HDF5_PLUGIN_PATH** for all filters but sz before loading **hd5plugin**: 

```python
def __clean_HDF5_PLUGIN_PATH():
    """
    if the libraries from hdf5plugin are in HDF5_PLUGIN_PATH, then remove them
    """
    import os
    import logging
    if "HDF5_PLUGIN_PATH" in os.environ:
        paths = os.environ["HDF5_PLUGIN_PATH"].split(":")
        keep = []
        for one_path in paths:
            if len(one_path) == 0:
                continue
            if 'h5z-sz' not in one_path:
                logging.info(f"removed {one_path} from HDF5_PLUGIN_PATH")
                continue
            keep.append(one_path)
        if len(keep) > 0:
            os.environ["HDF5_PLUGIN_PATH"] = ":".join(keep)
        else:
            del os.environ["HDF5_PLUGIN_PATH"]
```
Maybe a similar approach can be used.

**Comment by t20100** [comment]
I'd really prefer avoiding hidden modification of the environment variables when running `import hdf5plugin`.
Also, looking at libhdf5 source code on github, the `HDF5_PLUGIN_PATH` env var is read when initializing the lib and so it is too late when running `import hdf5plugin`.

PR #208 proposes another way to workaround this situation.

**Comment by t20100** [comment]
Closing as PR #208 provides a way to workaround this.

Please reopen if needed.

**Comment by juntyr** [comment]
> The problem is that for writing, some filters need to call some functions of `libhdf5` to get e.g., the data type or the item size, and so you need to link the filter with libhdf5.... This is an issue because the `libhdf5` that one use at runtime through a `h5py` wheel for example is not known at compile time of `hdf5plugin` + C symbols for Python extension are not loaded in the global namespace so not accessible to other packages.
> 
> To workaround this `hdf5plugin` provides stubs of `libhdf5` that are initialized at runtime and dynamically load C symbols from the `libhdf5` used by `h5py`.

I have been working on getting hdf5plugin to work inside Pyodide, and stumbled across this issue again (see https://lab.climet.eu/main/gist/github/juntyr/c08cec1105284c681fe3416e05ef2067/hdf5plugin.ipynb for a demo). While reading compressed files seems to mostly work, writing fails for at least the ZFP, SPERR, and SZ3 compressors but succeeds for LZ4 and Zstd

**Comment by t20100** [comment]
Some filters (like LZ4, Zstd, BZip2) do not need information about the data they compress (like datatype, size of the elements, chunk dimensions) do not need to call libhdf5 to compress/decompress.
Others (Blosc, bitshuffle, zfp..) need some information about the data and call libhdf5 to get those info.

Filters embedded in hdf5plugin are patched to replace static compile-time linking to libhdf5 with some dynamic loading of libhdf5 symbols with `dlopen`/`dlsym`.
To enable this, a [C init_filter](https://github.com/silx-kit/hdf5plugin/blob/0150a1a2337b2a473e2389155d8994b04a46fac1/lib/hdf5_dl.c#L184) function is added to the filters, and is called [here](https://github.com/silx-kit/hdf5plugin/blob/0150a1a2337b2a473e2389155d8994b04a46fac1/src/hdf5plugin/_utils.py#L153).
I don't know much about webassembly but I wouldn't be surprised if this approach does not work.

Can't you build the hdf5 filters independently and link them with libhdf5 like in https://github.com/usnistgov/libhdf5-wasm ? And install only the Python part of hdf5plugin (with `export HDF5PLUGIN_STRIP="all"`) ?

**Comment by juntyr** [comment]
I did some more digging and noticed that the plugin files, e.g. `plugins/libh5zfp.so`, don't export the `init_filter` symbol. It is exported by the intermediary `libhdf5_dl.a` build artefact, but for some reason, it is not exposed in the final shared library. I've been trying some things to change that but without any success so far

**Comment by juntyr** [comment]
I tried adding the hdf5_dl.c file directly to the sources of all plugins - this does succeed in getting the `init_filter` symbol exported, but it's obviously quite a hack

**Comment by t20100** [comment]
> it's obviously quite a hack

Yes, the embedded filters in this project relies on this ugly hack!
This was needed to allow linking with h5py wheel which embeds libhdf5.

The neat alternative is to link the filters against libhdf5 and place them in the folder where libhdf5 looks for filters.
Otherwise it would need libhdf5 to completely change the way the handle external filters....

**Comment by juntyr** [comment]
I made some more progress (some were bug fixes related to Pyodide) by skipping the init_filter workaround and directly dynamically linking with h5py's libhdf5.so (I can control the path and so set a rpath that links them up correctly). Now I can mostly load hdf5plugin and create datasets with compressors, though ZFP and SPERR are still acting up (see https://gist.github.com/juntyr/a9ce042d6303722f7444c87ae284190f). The ZFP error is especially weird since the `BIT_STREAM_WORD_TYPE=uint8` is set in `setup.py` and I haven't modified that. Any ideas?

**Comment by juntyr** [comment]
Btw these are the patches I'm currently applying in the build: https://github.com/climet-eu/pyodide-recipes/tree/climet-lab-0.4.0/packages/hdf5plugin/patches

**Comment by t20100** [comment]
> The ZFP error is especially weird since the BIT_STREAM_WORD_TYPE=uint8 is set in setup.py and I haven't modified that. Any ideas?

The ZFP lib is also used by the blosc2 filter and I don't think this one is built with `BIT_STREAM_WORD_TYPE=uint8`. Depending how this gets loaded, maybe you have an issue with duplicated conflicting symbols 🤷 

If you put the filters in the folder lihdf5 is looking for, you'd best not leave them inside the `hdf5plugin/plugins` folder and use `hdf5plugin` as a pure-Python project (this how it is packaged in Debian).
If building filters linked with libhdf5 (the best whener possible), you may consider building filters from https://github.com/HDFGroup/hdf5_plugins as done in https://github.com/usnistgov/libhdf5-wasm instead of using python build machinery.

**Comment by juntyr** [comment]
I don't think it's a runtime linking issue, but a compile-time one. But I have no idea where the build isolation breaks down. My quick fix was just to also enable the flag in blosc2. With that, both zfp and blosc2 work (and sperr also works, I just needed to give it >1d data)

**Comment by t20100** [comment]
> My quick fix was just to also enable the flag in blosc2

Are you sure it will be able to read blosc2/zfp compressed data compressed by another tool?

BTW, for testing, we have a few files that has been compressed by filters outside of this project: https://github.com/silx-kit/hdf5plugin/tree/main/test. Unfortunately there is no blosc2 file.
